File: solve/cpp/bogomolny.cc
#include "common.hh"

void recurse(char c)
{
    if (c == LAST)
    {
        output();
    }
    else
    { 
        for (int i = 0; i != n; ++i)
        {
            if (a[i] == 0)
            {
                a[i] = c;
                recurse(c + 1);
                a[i] = 0;
            }
        } 
    }
}

int main(int argc, char *argv[])
{
	if (init(argc, argv, UNORDERED) == 0)
    {
        fill(a, a + n, 0);
        recurse(FIRST); 
        return final(); 
    }
    return 1;
}

File: solve/lua/bogomolny.lua
#! /usr/bin/lua

require "common"

function recurse(k)
	for i = 1, n do
		if a[i] == 0 then
			a[i] = 96 + k
			if k == n then
				output()
			else
				recurse(k + 1)
			end
			a[i] = 0
		end
	end
end

skeleton(UNORDERED, function ()
	for i = 1, n do
		a[i] = 0
	end
	recurse(1)
end)
